### PR TITLE
Warn about redundant null cases in switch statements and expressions

### LINKED ISSUE #628: Warn on redundant null cases in switch expressions and enhanced switch statements
Take this example:

````
class SwitchRedundantNullCase {
    void foo(Object o) {
        // :: (nulltest.redundant)
        if (o == null) {
            System.out.println("o is null");
        }
        switch(o) {
        case Number n: System.out.println("Number: " + n); break;
        // :: (nulltest.redundant)
        case null: System.out.println("null"); break;
        default: System.out.println("anything else");
        }
    }
}
````

When we run `./checker/bin/javac -processor nullness SwitchRedundantNullCase.java -Alint=redundantNullComparison` we get:

````
RedundantSwitchOnNull.java:4: warning: [nulltest.redundant] redundant test against null; "o" is non-null
	  if (o == null) {
	        ^
1 warning
````

We should also get a warning about a `case null` when the expression is non-null.

### PR DESCRIPTION
**Pull request by Ao-senXiong**
Fixes #628.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 1]
```diff
@@ -767,6 +768,23 @@ public Void visitSwitch(SwitchTree tree, Void p) {
...
         }
+        if (checker.getLintOption(
+                NullnessChecker.LINT_REDUNDANTNULLCOMPARISON,
+                NullnessChecker.LINT_DEFAULT_REDUNDANTNULLCOMPARISON)) {
+            ExpressionTree expression = tree.getExpression();
+            List<? extends CaseTree> cases = tree.getCases();
+            AnnotatedTypeMirror switchType = atypeFactory.getAnnotatedType(expression);
+            for (CaseTree caseTree : cases) {
+                @SuppressWarnings("deprecation")
+                ExpressionTree caseExpression = caseTree.getExpression();
```
Comment (Ao-senXiong) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:779]: I used `getExpression()` because jkd 8, 11 do not have the methods `getExpressions()`. And the `getExpression()`  has been deprecated in jdk 17 and above.

Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:779]: Isn't it enough to use the `TreeUtils.hasNullCaseLabel(tree)` that's already computed for the previous `if`?
Just put that result into a local variable and reuse it here.

For helper methods for these version-specific getters, see `TreeUtilsAfterJava11`.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 2]
```diff
@@ -767,6 +768,23 @@ public Void visitSwitch(SwitchTree tree, Void p) {
         if (!TreeUtils.hasNullCaseLabel(tree)) {
             checkForNullability(tree.getExpression(), SWITCHING_NULLABLE);
         }
+        if (checker.getLintOption(
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:771]: Put the result of `getLintOption` into a field and init it in the constructor.

File: checker/tests/nullness/java21/FlowSwitch.java [hunk 3]
```diff
@@ -10,6 +10,7 @@ public class FlowSwitch {
   void test0(Number n) {
     String s = null;
     switch (n) {
+      // :: warning: (nulltest.redundant)
       case null, default: {
         // TODO: this should issue a dereference of nullable error.
```
Comment (wmdietl) [inline checker/tests/nullness/java21/FlowSwitch.java:15]: Note how this is inconsistent with the TODO here. Maybe this should be two test cases, one where the case is redundant and one where the TODO applies?

File: framework/tests/all-systems/java21/JEP441.java [hunk 4]
```diff
@@ -8,7 +8,7 @@
 // These are examples copied from:
 // https://openjdk.org/jeps/441
 
-@SuppressWarnings("i18n") // true postives.
+@SuppressWarnings({"i18n", "nullness:nulltest.redundant"}) // true postives.
```
Comment (wmdietl) [inline framework/tests/all-systems/java21/JEP441.java:11]: How about adding `@Nullable` annotations to the corresponding method parameters? That would make these additional valid tests for the new feature.

File: docs/CHANGELOG.md [hunk 5]
```diff
@@ -3,6 +3,9 @@ Version 3.40.0-eisop3 (November ??, 2023)
 
 **User-visible changes:**
 
+The nullness checker now issues redundant test against null warnings for switch statements when the expression is
```
Comment (wmdietl) [inline docs/CHANGELOG.md:6]: This is not just for switch statements, also for switch expressions. Please make sure there is or to add test coverage for both.
Try to simplify the sentence.

Comment (wmdietl) [inline docs/CHANGELOG.md:6]:
```suggestion
The Nullness Checker now issues redundant test against null warnings for switch statements when the expression is
```

File: docs/CHANGELOG.md [hunk 6]
```diff
@@ -3,6 +3,9 @@ Version 3.40.0-eisop3 (November ??, 2023)
...
+non-null and the case expression is null literal.
```
Comment (wmdietl) [inline docs/CHANGELOG.md:7]: This only happens when a command-line option is provided. How about pointing the reader to the section of the manual and/or listing the option?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for fixing this!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 1 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java]: > Isn't it enough to use the `TreeUtils.hasNullCaseLabel(tree)` that's already computed for the previous `if`? Just put that result into a local variable and reuse it here.
> 
> For helper methods for these version-specific getters, see `TreeUtilsAfterJava11`.

I think we still need logic here to locate which case should we report the error to.

File: docs/CHANGELOG.md (hunk 6 above)
Comment (Ao-senXiong) [reply docs/CHANGELOG.md]: ok Thanks

> This only happens when a command-line option is provided. How about pointing the reader to the section of the manual and/or listing the option?

Thanks! Addressed!

File: docs/CHANGELOG.md (hunk 5 above)
Comment (Ao-senXiong) [reply docs/CHANGELOG.md]: > This is not just for switch statements, also for switch expressions. Please make sure there is or to add test coverage for both. Try to simplify the sentence.

Thanks, I have updated it and please see whether you like it.

File: framework/tests/all-systems/java21/JEP441.java (hunk 4 above)
Comment (Ao-senXiong) [reply framework/tests/all-systems/java21/JEP441.java]: > How about adding `@Nullable` annotations to the corresponding method parameters? That would make these additional valid tests for the new feature.

Thanks! addressed!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 7]
```diff
@@ -776,9 +788,38 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+        if (lintredundantNullComparison) {
+            ExpressionTree expression = SwitchExpressionUtils.getExpression(switchExprTree);
+            List<? extends CaseTree> cases = SwitchExpressionUtils.getCases(switchExprTree);
+            checkSwitchNullRedundant(expression, cases);
+        }
         super.visitSwitchExpression17(switchExprTree);
     }
 
+    /**
+     * Reports an error if the switch expression is @NonNull and the case expression is null
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:800]:
```suggestion
     * Reports an error if the switch expression is {@code @NonNull} and the case expression is null
```

File: framework/tests/all-systems/java21/Issue6173.java [hunk 8]
```diff
@@ -4,10 +4,10 @@
 // @infer-jaifs-skip-test
 // @infer-ajava-skip-test
 // @infer-stubs-skip-test
-
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (wmdietl) [inline framework/tests/all-systems/java21/Issue6173.java:7]:
```suggestion
import org.checkerframework.checker.nullness.qual.Nullable;

```

File: framework/tests/all-systems/java21/JEP441.java [hunk 9]
```diff
@@ -8,7 +8,8 @@
 // These are examples copied from:
 // https://openjdk.org/jeps/441
 
-@SuppressWarnings("i18n") // true postives.
+import org.checkerframework.checker.nullness.qual.Nullable;
```
Comment (wmdietl) [inline framework/tests/all-systems/java21/JEP441.java:11]:
```suggestion
import org.checkerframework.checker.nullness.qual.Nullable;

```

File: docs/CHANGELOG.md [hunk 10]
```diff
@@ -3,6 +3,9 @@ Version 3.40.0-eisop3 (November ??, 2023)
 
 **User-visible changes:**
 
+The Nullness Checker now warns about unnecessary null checks in switch block when applying
+`-Alint=redundantNullComparison` command-line argument.
+
 **Implementation details:**
 
 **Closed issues:**
```
Comment (wmdietl) [inline docs/CHANGELOG.md:11]:
```suggestion
**Closed issues:**

eisop#628.
```

File: checker/tests/nullness/java21/NullRedundant.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            // :: warning: (nulltest.redundant)
+            case null: System.out.println("null"); break;
+            default: System.out.println("anything else");
+        }
+
+        var output = switch(o) {
+            case Number n -> "Number: " + n;
+            // :: warning: (nulltest.redundant)
+            case null -> "null";
+            default -> "anything else";
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:20]: Can you add another test that uses `case null, default:` and `case null, default ->`.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 12]
```diff
@@ -776,9 +788,38 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+     * @param cases the case expressions
+     */
+    private void checkSwitchNullRedundant(
+            ExpressionTree expression, List<? extends CaseTree> cases) {
+        AnnotatedTypeMirror switchType = atypeFactory.getAnnotatedType(expression);
+        for (CaseTree caseTree : cases) {
+            List<? extends ExpressionTree> caseExpressions =
+                    TreeUtilsAfterJava11.CaseUtils.getExpressions(caseTree);
+            for (ExpressionTree caseExpression : caseExpressions) {
+                if (caseExpression != null
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:813]: When is this null? Maybe for switch statements only do something for enhanced switch statements?
https://github.com/eisop/checker-framework/blob/d7f9dabf1240e40484ce2b3c1c6e9feecaabe29c/javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java#L2447

Suggestion above probably addresses this.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 13]
```diff
@@ -125,6 +127,9 @@ public class NullnessNoInitVisitor extends BaseTypeVisitor<NullnessNoInitAnnotat
     /** True if -AassumeAssertionsAreDisabled was passed on the command line. */
     private final boolean assumeAssertionsAreDisabled;
 
+    /** True if -Alint=permitRedundantNullComparison was passed on the command line. */
+    private final boolean lintredundantNullComparison;
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:131]: Or `lintRedundantNullComparison;` for proper camel casing.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 14]
```diff
@@ -767,6 +774,11 @@ public Void visitSwitch(SwitchTree tree, Void p) {
         if (!TreeUtils.hasNullCaseLabel(tree)) {
             checkForNullability(tree.getExpression(), SWITCHING_NULLABLE);
         }
+        if (lintredundantNullComparison) {
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:777]:
```suggestion
        if (lintredundantNullComparison && TreeUtils.isEnhancedSwitchStatement(tree)) {
```

Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:777]: Only enhanced switch statements can have a `null` case.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 7 above)
Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:800]: It's not that the "case expression is null"... one of the case expressions in the case trees is null. Try to formulate it a bit clearer.

File: checker/tests/nullness/java21/NullRedundant.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
+// @below-java21-jdk-skip-test
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:1]: Add a link to the issue.

File: checker/tests/nullness/java21/NullRedundant.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+class NullRedundant {
+    void test1(Object o) {
+        // :: warning: (nulltest.redundant)
+        if (o == null) {
+            System.out.println("o is null");
+        }
+        switch(o) {
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:9]:
```suggestion
        switch (o) {
```

File: checker/tests/nullness/java21/NullRedundant.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+            case Number n: System.out.println("Number: " + n); break;
+            // :: warning: (nulltest.redundant)
+            case null: System.out.println("null"); break;
+            default: System.out.println("anything else");
+        }
+
+        var output = switch(o) {
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:16]:
```suggestion
        var output = switch (o) {
```

File: checker/tests/nullness/java21/NullRedundant.java [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+            case Number n -> "Number: " + n;
+            // :: warning: (nulltest.redundant)
+            case null -> "null";
+            default -> "anything else";
+        };
+    }
+
+    Object test2(Object o) {
+        return switch(o) {
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:25]:
```suggestion
        return switch (o) {
```

File: checker/tests/nullness/java21/NullRedundant.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+            // :: warning: (nulltest.redundant)
+            case null -> "null";
+            default -> "anything else";
+        };
+    }
```
Comment (wmdietl) [inline checker/tests/nullness/java21/NullRedundant.java:30]: For completeness, add a test with a `@Nullable` parameter that shows you don't get a false positive.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 20]
```diff
@@ -776,9 +788,38 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+     * literal.
+     *
+     * @param expression the switch expression
+     * @param cases the case expressions
+     */
+    private void checkSwitchNullRedundant(
+            ExpressionTree expression, List<? extends CaseTree> cases) {
+        AnnotatedTypeMirror switchType = atypeFactory.getAnnotatedType(expression);
+        for (CaseTree caseTree : cases) {
+            List<? extends ExpressionTree> caseExpressions =
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:810]: This only iterates over the case expressions, whereas `isEnhancedSwitchStatement` only iterates over the case labels.
Which one is right? Or should both iterate over both the case expressions & labels?

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 21]
```diff
@@ -776,9 +788,38 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+                    TreeUtilsAfterJava11.CaseUtils.getExpressions(caseTree);
+            for (ExpressionTree caseExpression : caseExpressions) {
+                if (caseExpression != null
+                        && caseExpression.getKind() == Tree.Kind.NULL_LITERAL
+                        && switchType.hasEffectiveAnnotation(NONNULL)) {
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:815]: You could move this test out of the loop. If the type is not `@NonNull`, the method doesn't need to do anything and can return early.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks! This looks nice, but there are a few issues we need to discuss.

File: checker/tests/nullness/java21/NullRedundant.java (hunk 15 above)
Comment (Ao-senXiong) [reply checker/tests/nullness/java21/NullRedundant.java:1]: Sure

File: checker/tests/nullness/java21/NullRedundant.java (hunk 11 above)
Comment (Ao-senXiong) [reply checker/tests/nullness/java21/NullRedundant.java:20]: ok, I will add the test case

File: checker/tests/nullness/java21/NullRedundant.java (hunk 19 above)
Comment (Ao-senXiong) [reply checker/tests/nullness/java21/NullRedundant.java:30]: sure!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 21 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:815]: sure!

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 20 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:813]: According to java doc, they are pretty much the same except for the type(`getexpression` returns expression tree and `getlabels` returns tree. I have tested both and works ok. Let's use getlabels to keep consistent.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 12 above)
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java]: I will remove the checks here.

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 22]
```diff
@@ -776,9 +788,37 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+     * @param expression the switch expression
+     * @param cases the case expressions
+     */
+    private void checkSwitchNullRedundant(
+            ExpressionTree expression, List<? extends CaseTree> cases) {
+        AnnotatedTypeMirror switchType = atypeFactory.getAnnotatedType(expression);
+        if (!switchType.hasEffectiveAnnotation(NONNULL)) {
+            return;
+        }
+        for (CaseTree caseTree : cases) {
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:812]:
```suggestion
        outer: for (CaseTree caseTree : cases) {
```

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java [hunk 23]
```diff
@@ -776,9 +788,37 @@ public void visitSwitchExpression17(Tree switchExprTree) {
...
+            List<? extends Tree> caseLabels = TreeUtilsAfterJava11.CaseUtils.getLabels(caseTree);
+            for (Tree caseLabel : caseLabels) {
+                if (caseLabel.getKind() == Tree.Kind.NULL_LITERAL) {
+                    checker.reportWarning(caseLabel, "nulltest.redundant", expression.toString());
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:816]: There can only be one null case, so we can stop once we found it.

**Review by wmdietl — COMMENTED** [review]
A few final touches...

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (Ao-senXiong) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessNoInitVisitor.java:816]: > There can only be one null case, so we can stop once we found it.

Thanks!

**Review by wmdietl — APPROVED** [review]
Looks good, thanks!

File: docs/CHANGELOG.md [hunk 24]
```diff
@@ -3,10 +3,15 @@ Version 3.40.0-eisop3 (November ??, 2023)
...
 **User-visible changes:**
 
+The Nullness Checker now warns about redundant null cases in switch statements and expressions when
+using the `-Alint=redundantNullComparison` command-line argument.
+
 **Implementation details:**
 
 **Closed issues:**
 
+eisop#628.
```
[the remark this answers is not in the document]
Comment (Ao-senXiong) [reply docs/CHANGELOG.md:13]: Addressed this comment in [9b392c4](https://github.com/eisop/checker-framework/pull/618/commits/9b392c43b461c04aa60f9132aa185394ce1574cb).
